Mathlib/Algebra/Star/NonUnitalSubsemiring.lean

```diff
@@ -0,0 +1,233 @@
+/-
+Copyright (c) 2024 Christopher Hoskin. All rights reserved.
+Released under Apache 2. license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Subsemigroup.Basic
+import Mathlib.RingTheory.NonUnitalSubsemiring.Basic
+import Mathlib.Algebra.Ring.Subsemiring.Basic
+import Mathlib.Algebra.Star.Center
+
+/-!
+# Non-unital Star Subsemirings
+
+In this file we define `NonUnitalStarSubsemiring`s and the usual operations on them.
+
+## Implementation
+
+This file is heavily inspired by `Mathlib.Algebra.Star.NonUnitalSubalgebra`.
+
+-/
+
+universe v w w'
+
+variable {A : Type v} {B : Type w} {C : Type w'}
+
+/-- A non-unital star subset is a subset which is closed under the `star`
+operation. -/
+structure StarSubset (A : Type v) [Star A] : Type v where
+  /-- The underlying set of a `SubStar`. -/
+  carrier : Set A
+  /-- The `carrier` of a `StarSubset` is closed under the `star` operation. -/
+  star_mem' : ∀ {a : A} (_ha : a ∈ carrier), star a ∈ carrier
+
+/-- A sub star magma is a subset of a magma which is closed under the `star`-/
+structure SubStarmagma (M : Type v) [Mul M] [Star M] extends Subsemigroup M, StarSubset M : Type v
```

Comment:
At the very least, the capitalization is wrong, but also, I think we should match naming with `Subsemigroup`. If we want to switch the naming of both to `Magma`, then fine, but that should be a separate PR.
```suggestion
/-- A sub star magma is a subset of a magma which is closed under the `star`-/
structure SubStarSemigroup (M : Type v) [Mul M] [Star M] extends Subsemigroup M, StarSubset M : Type v
```

Reply:
Possibly I should have used  `NonAssocSemigroup` in #7060 instead of `Magma` for consistency with `(NonUnital)NonAssoc(Semi)ring`?